src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -1,67 +1,87 @@
 internal sealed class CommandWriter : IAsyncDisposable
 {
+    private readonly ILogger<CommandWriter> _logger;
+    private readonly ObjectPool _pool;
+    private readonly object _lock = new();
+    private readonly CancellationTokenSource _cts;
     private readonly ConnectionStatsCounter _counter;
     private readonly TimeSpan _defaultCommandTimeout;
     private readonly Action<PingCommand> _enqueuePing;
     private readonly NatsOpts _opts;
-    private readonly PipeWriter _pipeWriter;
     private readonly ProtocolWriter _protocolWriter;
-    private readonly ChannelWriter<QueuedCommand> _queuedCommandsWriter;
-    private readonly SemaphoreSlim _semLock;
-    private Task? _flushTask;
+    private readonly Task _readerLoopTask;
+    private readonly HeaderWriter _headerWriter;
+    private readonly Channel<int> _channelLock;
+    private ISocketConnection? _socketConnection;
+    private PipeReader? _pipeReader;
+    private PipeWriter? _pipeWriter;
     private bool _disposed;
 
-    public CommandWriter(NatsOpts opts, ConnectionStatsCounter counter, Action<PingCommand> enqueuePing, TimeSpan? overrideCommandTimeout = default)
+    public CommandWriter(ObjectPool pool, NatsOpts opts, ConnectionStatsCounter counter, Action<PingCommand> enqueuePing, TimeSpan? overrideCommandTimeout = default)
     {
+        _logger = opts.LoggerFactory.CreateLogger<CommandWriter>();
+        _pool = pool;
         _counter = counter;
         _defaultCommandTimeout = overrideCommandTimeout ?? opts.CommandTimeout;
         _enqueuePing = enqueuePing;
         _opts = opts;
-        var pipe = new Pipe(new PipeOptions(
-            pauseWriterThreshold: opts.WriterBufferSize, // flush will block after hitting
-            resumeWriterThreshold: opts.WriterBufferSize / 2,  // will start flushing again after catching up
-            minimumSegmentSize: 16384, // segment that is part of an uninterrupted payload can be sent using socket.send
-            useSynchronizationContext: false));
-        PipeReader = pipe.Reader;
-        _pipeWriter = pipe.Writer;
-        _protocolWriter = new ProtocolWriter(_pipeWriter, opts.SubjectEncoding, opts.HeaderEncoding);
-        var channel = Channel.CreateUnbounded<QueuedCommand>(new UnboundedChannelOptions { SingleWriter = true, SingleReader = true });
-        _semLock = new SemaphoreSlim(1);
-        QueuedCommandsReader = channel.Reader;
-        _queuedCommandsWriter = channel.Writer;
+        _protocolWriter = new ProtocolWriter(opts.SubjectEncoding);
+        _channelLock = Channel.CreateBounded<int>(1);
+        _headerWriter = new HeaderWriter(_opts.HeaderEncoding);
+        _cts = new CancellationTokenSource();
+        _readerLoopTask = Task.Run(ReaderLoopAsync);
     }
 
-    public PipeReader PipeReader { get; }
-
-    public ChannelReader<QueuedCommand> QueuedCommandsReader { get; }
+    public void Reset(ISocketConnection? socketConnection)
+    {
+        var pipe = new Pipe(new PipeOptions(
```

Comment:
The current implementation does not reset the Pipe when a socket connects.  The reason for this is that a command may already be in the Pipe, but not sent.

This was OK - since the Reader would mark data as Examined but not Processed, the next socket would read from the same Pipe, and pick up that command that was still in the Pipe but had never been sent.

By resetting the pipe here, I think we are potentially losing that unsent command?

Reply:
It is bit of a pain swapping out reader and writer as well. In that case to avoid data loss, additionally I guess we should set buffer position for read advance after socket send?

Reply:
Lots of the logic in `NatsPipeliningWriteProtocolProcessor` dealt with figuring out command boundaries, and only marking data as Consumed in the Pipe Reader after the entire command was sent.  Would need to revive some of that logic - maintain an Unbounded Channel of Command Sizes that have been Flushed to the Pipe Reader, and Peek / Read those after each successful `Send` to figure out how much data can be marked as Consumed

Reply:
need to think about it a bit but my first thought is it would not matter because writer flush and reader read are always in sync but that would mean getting rid of the sendMemory limitation which might not be a bad idea anyway. write buffer size (I think set to 64K at the moment - which is pipelines default) would be the largest we write to socket.

Reply:
> need to think about it a bit but my first thought is it would not matter because writer flush and reader read are always in sync but that would mean getting rid of the sendMemory limitation which might not be a bad idea anyway. write buffer size (I think set to 64K at the moment - which is pipelines default) would be the largest we write to socket.

Two questions, semi related:

1: Do we have tests and/or benchmarks to test differences and functionality for larger payloads? I know it's... far from ideal but sometimes it happens and it may help us see opportuntities for improvement
2: Per comment on buffer size, should we have a follow up to add configurability?

Reply:
The main issue I am trying to highlight is a partially sent command. For example, a 5K publish. The first 2K is sent, then the socket drops. That 2K needs to be re-sent next time the socket connects, and the remaining 3K also needs to be sent.

So Bytes shouldn't be marked as consumed right after they are sent. But once an entire command has been sent, its bytes can be marked as consumed. So the reader needs knowledge of the commands sizes ahead of time.

Reply:
> writer flush and reader read are always in sync

Socket Send doesn't always send the full buffer, it returns the number of bytes actually sent, which could be less than the supplied buffer.

Plus large commands could be bigger than the Flush size, and Flush is called once but multiple Reads need to happen.

Reply:
ok I get it. This is the complication I wanted to avoid hence dropping the buffer on reconnect. I think data loss may be acceptable since we're in at most once territory. Also some messages might fail on a new connection.

Reply:
I don't know if dropping is preferable if we can help it. I get that we are in at-most once territory but we also say that sent commands means accepted to a send buffer, and that we transparently handle reconnects. In my mind that means that a reconnect should not cause a command in the send buffer to be dropped during a reconnect.

Reply:
> Also some messages might fail on a new connection

I think you're right, maybe PING and SUB should be skipped if the socket cycles?  It definitely complicates the logic...

Reply:
one thing we should do though that we should drop the message from the buffer whether the socket sent was successful or not to ensure at-most-once. I'm not sure but there may be cases part of the data is sent across even though send threw an exception (might be totally wrong here).

Reply:
I added a test in #360 to capture the current behavior, should we want to preserve that